### ISSUE TITLE
[FEATURE REQUEST] Bank Select LSB support with wBank

### ISSUE THREAD

**Issue by sylvia-leaf** [SFe-Team-was-taken/SFeReferenceImplementation_Core#3]
**Is your feature request related to a problem? Please describe.**
Spessasynth is limited to bank select MSB. Bank select LSB is unsupported, as with other soundfont players.

**Is your feature related to the sound library or the app/website?**
App/website 

**Describe the solution you'd like**
Implement bank select LSB using SFe specification method.

**Describe alternatives you've considered**
SpessaSynth currently includes a few "hacks" to get XG MIDI files to play properly when it detects an XG reset. However, these hacks will never be a workable substitute for proper support for two bank selects.

**Additional context**
Support for two bank selects is a foundational feature of SFe; it allows XG to work properly.

**Comment by sylvia-leaf** [comment]
@spessasus Don't worry about it!

> Firstly, can you transfer this issue to the core repo?

I've transferred all issues to the core repository. All future issues will be in the appropriate repository.

---

> Regarding the patch you've, I think you've missed the DLS and sf2 writing functions.

This will come after everything else. I want to make sure that SFe data can be read properly before even attempting to write functions to write. 

---

Status of each location to make changes:

- [ ] XG hacks
  - A lot of things depend on this.
  - Will likely detour these functions until SFe detection implemented
  
- [x] SF manager
	- Done
    - Still need to rename stuff

- [ ] MIDI channel
	- Not started

- [ ] isBankLSB
	- Not started

- [ ] \[Challenge\] RMIDI editor
	- Not started
    - After read support

- [ ] \[Challenge\] RMIDI writer
	- Not started
    - After read support

- [ ] \[Challenge\] Used keys for presets
	- This is in progress
    - Depends on XG hacks

---

> The hard part will be converting SF2's into SFe in such a way that the hacks are "baked in" i.e., lsb and msb will be processed normally, it's up to the bank to behave correctly.

So, what we want to do is to be able to load SF2 and SFe simultaneously, and only have the XG hacks apply to SF2?

If so, then what is possibly the hardest thing to do is to determine how conflicts between LSB-remapped SF2's and LSB-native SFe files are resolved. What would be the best solution for this problem?

---

>Regarding the UI, you're mostly interested in `synthui_selector` for selecting presets

This will be tracked by a new issue in the `SFeReferenceImplementation_WebApp` repository.

---

>And there's one question: The bank offset functionality in the manager. I propose keeping it as MSB-only since I see no reason to add it to bank LSB, as it's only used in sf2 environments like AWE32 MIDIs and RMIDI files.

I'll add bank offset functionality to LSB at a later time, only when absolutely everything else is completed.

**Comment by sylvia-leaf** [comment]
The preset reader can now detect an SFe bank. However, I still need to know how I can detect the type of MIDI file from the preset reader, so I can copy the presets to the correct locations with XG hacks.

**Comment by spessasus** [comment]
You can't do that and I do not understand why this is needed.

You can't because:
- The synth can exist without a sequencer.
- Copying presets every time an XG on is received is not a very good idea.
- Synth should not modify the banks themselves.
- Adding stacked banks to the mix using soundfont manager will make this a big mess.

When I meant duplicating the presets, I meant doing it always, during SF2 load time. There's no reason why we would care about the current system, as the solution I proposed works for both:
- untouched MSB + LSB 0 for GS, works just like earlier. Even if an LSB is specified, the fallback mechanism will default to LSB 0 if a matching MSB is found
- MSB 0 + LSB copied: XG sets instrument variation with LSB, so the behavior is correct.

If you implement that, I'll be sure to extensively test it with my MIDI files to see how this solution works and we can finetune it from there.


Regarding the loading system:
- The presets are loaded during parsing, however samples and per key:velocity generators, modulators and samples are loaded per note-on request, then cached. So the code applying the generator additions, identical modulators and ranges is ran only once.
- The cache can be cleared either via synth.clearCache() or for every soundfont manager operation.
- The sequencer has an unique ability of going through the MIDI file to find all used presets and key:velocity combinations, then preloading all zones and samples for it:

https://github.com/spessasus/spessasynth_core/blob/1f36fef686b7a18f1e6d9cf9de0684074913b856/src/sequencer/song_control.js#L71-L88

**Comment by sylvia-leaf** [comment]
@spessasus I understand what you mean.

Do you want even the XG drums to be copied, or would you want them to be left alone?

**Comment by spessasus** [comment]
> If bank is: 0, 64, 120, 126, 127 or 128 (drum) keep as is.

By bank I mean the SF2 bank number. By keep as is, I mean keeping the bank to MSB and leaving LSB at 0.

 If you want to know why these specific numbers, the answer is:
- 0: pretty obvious, zero is zero.
- 64: XG bank MSB 64 classifies as "SFX", so leave as is.
- 120: GM2 drums: use MSB so leave as is.
- 126: XG SFX drums: same as above.
- 127: XG regular drums: same as above.
- 127: SF2 drums, duh!

Now that I'm thinking about it, maybe we could do one thing:
if there are no presets on bank MSB 127, copy presets from bank 128 there. It'll work 99% of the time IMO, as if you're using a MIDI that specifies bank 127 in GS, you're probably using a soundfont that has MT-32 presets on 127, which do not get overriden.

So that's the logic for now. As I've mentioned before, let's stick with that for now and we can always adjust it as we go. No pressure, we aren't under a time limit :-)

**Comment by sylvia-leaf** [comment]
@spessasus I've added preset duplication as you've described:

- All MSB values except for reserved XG/GM2 values are now duplicated as LSB if SF version is detected as legacy 2.0x
- If there are no MSB=127 presets, the drums are duplicated there.
- Additionally, MSB=120 is duplicated for GM2.

Please test this. Now that this is implemented, do we need the `parseBankSelect` or `chooseBank` functions? If not, then this will be worked on next.

**Comment by sylvia-leaf** [comment]
@spessasus I've adapted the XG hacks code so MSB is for the most part passed through, with the exception of drums, which require special handling.

However, I am a bit confused about the `programchange` call event, and what the bank select is used for there. I understand the `programChange` function but this calls a `callEvent` which I find challenging to understand.

Do you have any insights about what I need to do, or whether I even need to do anything?

**Comment by spessasus** [comment]
> I've adapted the XG hacks code so MSB is for the most part passed through, with the exception of drums, which require special handling.

Given that drums are automatically duplicated to 127, I don't think any special handling is necessary here.

Regarding program change, it calls
```js
this.sentBank = Math.min(128, preset.bank + p.bankOffset);
```
so it sends the soundfont bank number of the preset + bank offset applied by the soundfont manager, which is returned from `getPreset`:

https://github.com/spessasus/spessasynth_core/blob/1f36fef686b7a18f1e6d9cf9de0684074913b856/src/synthetizer/audio_engine/engine_components/soundfont_manager.js#L175-L181

This is also why I propose to leave bank offset to MSB alone, I don't see any practical reason for adding it to LSB.

CallEvent is used for the [event system](https://github.com/spessasus/spessasynth_core/wiki/Synth-Event-Types), which for example tells the synth controller or keyboard that note was pressed or a controller was changed.

**Comment by sylvia-leaf** [comment]
@spessasus Sorry for the delay, but thank you for reporting the bug!

That being said, I am completely unsure how I could test everything. I've got many questions about all of this.

Firstly, I checked the soundfont you linked and I didn't see any XG drums. Is your branch of SpessaSynth automatically remapping drums or something? If so, then I will have to implement this (which may be difficult).

The next question is if a step-by-step guide can be provided to set up the example so I could test my implementation against your branch. If this is possible, then I would be grateful!

With that out of the way, I think that the bug is caused by my implementation of copying presets to banks. I'll keep working on it.

**Comment by sylvia-leaf** [comment]
@spessasus It is fully implemented in the SpessaFont fork, however I've found a strange issue that I don't know how to solve when trying to add full support in the SpessaSynth web app.

I've widened most or all of the code to add LSB bank support, however the set_event_listeners.js still uses MSB-only banks for some reason. Because I don't fully know how this "event handler" works, it's unclear why the data that the event handler is receiving remains MSB-only.

What will I need to do to add LSB support to this piece of code that I don't know about? Is it in spessasynth_core?

**Comment by sylvia-leaf** [comment]
I've made sure that bankLSB is included, but it still passes MSB only. I don't have any clue what is going wrong. If you look [here](https://github.com/SFe-Team-was-taken/SFeReferenceImplementation_Core/blob/af28bf11f4e4af1a57f73c6c2135e3f06cd76204/src/synthetizer/audio_engine/engine_methods/program_change.js) then you can clearly see that bankLSB is there.

**Comment by spessasus** [comment]
No, you didn't.

https://github.com/SFe-Team-was-taken/SFeReferenceImplementation_Core/blob/af28bf11f4e4af1a57f73c6c2135e3f06cd76204/src/synthetizer/audio_engine/engine_methods/controller_control/reset_controllers.js#L75-L79

A simple ctrl+shift+f search for "programchange" showed that you indeed forgot to change it.

**Comment by spessasus** [comment]
PS: Why is the `build` package installed in all SFe's reference repos? It's not needed and node says that it has vulnerabilities.

**Comment by sylvia-leaf** [comment]
@spessasus Sorry about that! I didn't catch everything, thank you for pointing out where I forgot to change something. Now it works.

However, there are still many components in core that are MSB-only. But all of the components that allow SpessaSynth's web app to function properly have been changed to take LSB into account.

Now, I will be quite busy with studying relatively soon, but after this I'm probably going to work on adding wav support, because the SFe specification requires this.

---

- [x] Basic support
- [ ] MIDI editor
- [ ] Sequencer
- [ ] Soundfont loading (done?)
- [ ] Synthetizer
- [ ] Utils (done?)

---

Talking about build always getting installed, I think it may be a side effect of running `npm install build` to install all dependencies for the build to function. Maybe this is not required.

**Comment by spessasus** [comment]
> Talking about build always getting installed, I think it may be a side effect of running npm install build to install all dependencies for the build to function. Maybe this is not required.

I'm sorry, what? To install all dependencies you just do `npm install`, not `npm install build`. `npm install build` is the same as `npm install react`, for example. `build` is just some npm package.

**Comment by sylvia-leaf** [comment]
Honestly, it is me being silly. I apologise.

**Comment by spessasus** [comment]
> I've found a bug with the current reference implementation master.
> 
> [On 'Til Morning by David Reading.zip](https://github.com/user-attachments/files/20850383/On.Til.Morning.by.David.Reading.zip)
> 
> with [this soundfont](https://musical-artifacts.com/artifacts/2592) causes audibly incorrect drums.
> 
> Run the midi_player_node example on both my branch and SFe's and you can hear that SFe uses GS's analog while core uses XG's analog as it should be. Note that you need to `npm i speaker` for the examples to work.

This bug is still present, any updates?

**Comment by spessasus** [comment]
LSB support has been implemented into upstream.

I even included SFe LSB read support :-) If detected, it won't do LSB shadowing.

**Comment by spessasus** [comment]
@sylvia-leaf

All you have to do (I think) is add write support for LSB and this is done. I did the hard part which is the preset logic and MIDI editor and RMIDI writers :D

**Comment by sylvia-leaf** [comment]
@spessasus I am still struggling to build SpessaSynth 4.0, because I get an "ERR_UNKNOWN_FILE_EXTENSION" when trying to build spessasynth_lib. 

It seems that the build script does not like typescript files, I don't know what I am doing wrong.

**Comment by sylvia-leaf** [comment]
@spessasus

```
PS E:\sfe\ReferenceImplementation\spessasynth_lib> npm run build

> spessasynth_lib@4.0.0 build
> npm run test:types && npm run local-build && npm run build:pages


> spessasynth_lib@4.0.0 test:types
> eslint src/ && tsc -b


> spessasynth_lib@4.0.0 local-build
> npm run build:npm && npm pack


> spessasynth_lib@4.0.0 build:npm
> node build_scripts/build_npm.ts

node:internal/modules/esm/get_format:218
  throw new ERR_UNKNOWN_FILE_EXTENSION(ext, filepath);
        ^

TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for E:\sfe\ReferenceImplementation\spessasynth_lib\build_scripts\build_npm.ts
    at Object.getFileProtocolModuleFormat [as file:] (node:internal/modules/esm/get_format:218:9)
    at defaultGetFormat (node:internal/modules/esm/get_format:244:36)
    at defaultLoad (node:internal/modules/esm/load:122:22)
    at async ModuleLoader.loadAndTranslate (node:internal/modules/esm/loader:479:32)
    at async ModuleJob._link (node:internal/modules/esm/module_job:112:19) {
  code: 'ERR_UNKNOWN_FILE_EXTENSION'
}

Node.js v22.11.0
```

**Comment by spessasus** [comment]
It looks like you have an older version of node. Node started supporting ts files natively only recently.

Try running `npx tsx build_scripts/build_npm.ts` instead.

If it works, I'll tweak the build script soon.

**Comment by spessasus** [comment]
I have freshly cloned the repositories to check, and just doing :

`npm install --verbose
npm run local-build`

in core, then lib works fine.

I have no idea why dts fails in yours. Maybe try replacing `Float32Array` type in these with `Float32Array<ArrayBuffer>` explicitly?

**Comment by stgiga** [comment]
BWTC32Key uses Uint8Arrays to handle binary data.

**Comment by spessasus** [comment]
Note that I have removed the preset shadowing with LSB as the bank selection algorithm attempts to use MSB as LSB if LSB is not available.
